### PR TITLE
Tokenize NavigationBar

### PR DESCRIPTION
**Pull request by amgleitman**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

This is a basic tokenization of NavigationBar.

The scope for this includes tokenizing elements of a basic navigation bar, including title colors and fonts.

As of aa9640c, this change does not yet apply tokenization to the accessory view (e.g., a pill button bar or a search bar) because each component has its own way of representing its style, and that logic doesn't necessarily belong in NavigationBar itself. We're still thinking about the best way to implement this.

### Verification

Verified using the test page. These screenshots show the token overrides in place.

One possible concern is that the light-mode override colors look weird on a brand background, but this is more of a pedagogical example to show that we can indeed override these tokens.

<details>
<summary>Visual Verification</summary>

| Example 1 | Example 2 | Example 3 | Example 4 |
|-|-|-|-|
| Simulator Screen Shot - iPhone 14 Pro - 2023-03-31 at 11 44 03 | Simulator Screen Shot - iPhone 14 Pro - 2023-03-31 at 11 44 16 | Simulator Screen Shot - iPhone 14 Pro - 2023-03-31 at 11 44 50 | Simulator Screen Shot - iPhone 14 Pro - 2023-03-31 at 11 45 03 |
</details>

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1684)

### CODE REVIEW

File: ios/FluentUI/Navigation/NavigationBar.swift [hunk 1]
```diff
@@ -107,43 +107,11 @@ open class NavigationBar: UINavigationBar, TokenizedControlInternal, TwoLineTitl
...
-                return defaultColor
-            case .system:
-                return fluentTheme.color(.background3)
-            case .custom:
-                return customColor ?? defaultColor
-            }
-        }
     }
 
+    @available(*, deprecated, renamed: "backgroundColor(forStyle:theme:)", message: "NavigationBar has been tokenized. Use the replacement method, passing in .system for the style.")
```
Comment (edjamesmsft) [inline ios/FluentUI/Navigation/NavigationBar.swift:112]: Unfortunately, since control tokens cannot be used in Objective C, and we don't have a plan for when they will be available, I don't think we can deprecate this API yet.

File: ios/FluentUI/Navigation/NavigationBar.swift [hunk 2]
```diff
@@ -301,7 +271,7 @@ open class NavigationBar: UINavigationBar, TokenizedControlInternal, TwoLineTitl
         }
     }
 
-    var titleView = LargeTitleView() {
+    var titleView = AvatarTitleView() {
```
Comment (edjamesmsft) [inline ios/FluentUI/Navigation/NavigationBar.swift:274]: This change seems unrelated to tokenizing. Should this be done in a separate PR? Additionally, what is the rationale for moving from LargeTitleView to AvatarTitleView?

File: ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift [hunk 3]
```diff
@@ -53,6 +53,22 @@ public class ControlTokenSet<T: TokenSetKey>: ObservableObject {
...
     }
 
+    /// Convenience method to override multiple tokens from another token set.
+    ///
+    /// This is useful if `otherTokenSet` belongs to a parent control and `self` belongs to a child control.
+    ///
+    /// - Parameter otherTokenSet: The token set we will be pulling values from.
+    /// - Parameter mapping: A `Dictionary` that maps our own tokens that we wish to override with
+    /// their corresponding tokens in `otherTokenSet`.
+    public func setOverrides<U>(from otherTokenSet: ControlTokenSet<U>, mapping: [T: U]) {
```
Comment (laminesm) [inline ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift:63]: Is this a safe addition? With our current token philosophy we want to add subcontrol tokens in the parent control and any overrides should happen from the parent. Wouldn't this open us up to potentially strange behaviors in situations where overrides occur from both parent and child?

File: ios/FluentUI/Navigation/NavigationBarTokenSet.swift [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                return .uiColor {
+                    switch style() {
+                    case .primary, .default, .custom:
+                        return UIColor(light: theme.color(.brandBackground1).light,
+                                       dark: theme.color(.background3).dark)
+                    case .system:
+                        return theme.color(.background3)
+                    }
+                }
+            case .buttonTintColor, .subtitleColor: // By default, these return the same color
```
Comment (edjamesmsft) [inline ios/FluentUI/Navigation/NavigationBarTokenSet.swift:46]: Nit: I don't think this comment is necessary.

File: ios/FluentUI/Navigation/NavigationBar.swift [hunk 5]
```diff
@@ -107,43 +107,11 @@ open class NavigationBar: UINavigationBar, TokenizedControlInternal, TwoLineTitl
         case primary
```
Comment (laminesm) [inline ios/FluentUI/Navigation/NavigationBar.swift:107]: Shouldn't `NavigationBar.Style` live in `NavigationBarTokenSet`?

File: ios/FluentUI/Navigation/NavigationBar.swift [hunk 6]
```diff
@@ -107,43 +107,11 @@ open class NavigationBar: UINavigationBar, TokenizedControlInternal, TwoLineTitl
...
-            case .system:
-                return fluentTheme.color(.background3)
-            case .custom:
-                return customColor ?? defaultColor
-            }
-        }
     }
 
+    @available(*, deprecated, renamed: "backgroundColor(forStyle:theme:)", message: "NavigationBar has been tokenized. Use the replacement method, passing in .system for the style.")
     @objc public static func navigationBarBackgroundColor(fluentTheme: FluentTheme) -> UIColor {
```
Comment (laminesm) [inline ios/FluentUI/Navigation/NavigationBar.swift:113]: Could you make this an an optional `FluentTheme` and use the shared theme as fallback (#1673 for reference)

File: ios/FluentUI/Navigation/Views/AvatarTitleView.swift [hunk 7]
```diff
@@ -5,21 +5,19 @@
 
 import UIKit
 
-// MARK: LargeTitleView
```
Comment (edjamesmsft) [inline ios/FluentUI/Navigation/Views/AvatarTitleView.swift:8]: Ohhh I see now. Given the scope of this change, it might be good to have broken this into two separate PRs. Easier to review, easier to roll back if something goes sideways.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (laminesm) [inline ios/FluentUI/Navigation/NavigationBar.swift:112]: Could you please clarify what this message is for? This function is primarily for our objc callers and seems relatively unchanged.

File: ios/FluentUI/Navigation/NavigationBarTokenSet.swift (hunk 4 above)
Comment (laminesm) [inline ios/FluentUI/Navigation/NavigationBarTokenSet.swift:46]: nit: remove redundant comment

File: ios/FluentUI/Navigation/Views/AvatarTitleView.swift [hunk 8]
```diff
@@ -158,18 +147,32 @@ class LargeTitleView: UIView, TwoLineTitleViewDelegate {
         setupAccessibility()
         twoLineTitleView.delegate = self
 
-        NotificationCenter.default.addObserver(self,
-                                               selector: #selector(themeDidChange),
-                                               name: .didChangeTheme,
-                                               object: nil)
+        tokenSet.registerOnUpdate(for: self) { [weak self] in
+            self?.updateAppearance()
```
Comment (edjamesmsft) [inline ios/FluentUI/Navigation/Views/AvatarTitleView.swift:151]: Above, you setTitleColor after updateAppearance, but here you don't setTitleColor. That to me means either setTitleColor should be moved into updateAppearance, or also called here (preferably the former).

File: ios/FluentUI/Navigation/NavigationBarTokenSet.swift (hunk 4 above)
Comment (laminesm) [reply ios/FluentUI/Navigation/NavigationBarTokenSet.swift:46]: Just saw this was mentioned already by @edjamesmsft so I'm gonna resolve this

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 2 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:274]: `LargeTitleView` is a bit of a misnomer now since it can also host a `TwoLineTitleView`. The defining characteristics of this view are that (a) it can show an avatar, and (b) its titles are leading-aligned. I chose the first one since the avatar is definitely an attention-grabber in practice.

It's tangentially related since we're tokenizing this component, so I think now's a fine time to rename it.

File: ios/FluentUI/Navigation/Shy Header/ShyHeaderController.swift [hunk 9]
```diff
@@ -254,8 +254,8 @@ class ShyHeaderController: UIViewController {
             }
         }
 
-        // if the originator is a LargeTitleView, make sure it belongs to this heirarchy
```
[the remark this answers is not in the document]
Comment (amgleitman) [reply ios/FluentUI/Navigation/Shy Header/ShyHeaderController.swift:257]: See [this comment](https://github.com/microsoft/fluentui-apple/pull/1684#discussion_r1154847084).

File: ios/FluentUI/Navigation/NavigationBarTokenSet.swift (hunk 4 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBarTokenSet.swift:46]: I don't think it hurts to keep it.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: Since this method is publicly exposed, it's possible that a client might be calling this. This is for their benefit so they explicitly know we're deprecating this instead of just getting faced with a random compiler error.

Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: Why can't we deprecate this? Even if we can't use control tokens in Objective-C, `NavigationBar.Style` and `FluentTheme` should be fair game.

Comment (laminesm) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: I guess I just don't understand why the deprecation message is placed on the `navigationBarBackgroundColor` function and not the `backgroundColor(forStyle:theme:)` function.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 6 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift]: Ooh, good idea! Done with a7a741b.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: `navigationBarBackgroundColor` is the old function, while `backgroundColor(forStyle:theme:)` is the new one.

File: ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift (hunk 3 above)
Comment (amgleitman) [reply ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift:63]: Do we have a preferred way to pass tokens to child controls? I don't see any examples of this currently in this repo.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (laminesm) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: I see. I don't think we should be attempting to replace this function with `backgroundColor(forStyle:theme:)`. We always want to simplify our APIs for our objc clients, and this new function does the opposite by adding a new parameter to the call.

Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: I thought that we wanted to move all functionality regarding appearance to `NavigationBarTokenSet`. Is this not the case anymore?

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 5 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:107]: The `Style` enum doesn't contain any information inherently about the token values themselves, so I think it's fine where it is now. Let the token set handle these translations.

File: ios/FluentUI/Navigation/Views/AvatarTitleView.swift (hunk 8 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/Views/AvatarTitleView.swift:151]: Upon closer inspection, `updateAppearance` already calls `setTitleColor`, so we can take a third option of removing the extra call.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: Previously, the API would assume that we're using a `NavigationBar` with the `.system` style, which is not a very good assumption to make IMHO. At least we tell our clients to make sure that they absolutely want this style.

Comment (edjamesmsft) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: We want to avoid breaking API changes wherever possible. Is this a necessary change? Is it necessary right now?

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 2 above)
Comment (edjamesmsft) [reply ios/FluentUI/Navigation/NavigationBar.swift:274]: Just to clarify - is it DIRECTLY related, or TANGENTALLY related? If it's tangental, as in, not very related, I really would rather it be in a separate PR.

In general, keeping PRs small and targeted is a good choice. Especially with a PR as big as this one has become, it might have been good to do this rename separately, after the large functionality change.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: Leaving it in is an unnecessary accumulation of debt. I'd rather nip it in the bud now than put it off for later, especially when the resulting cost for clients is very low (we're LITERALLY telling them how to update to the new API).

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 2 above)
Comment (amgleitman) [reply ios/FluentUI/Navigation/NavigationBar.swift:274]: It's relevant here because of the addition of `AvatarTitleViewTokenSet`, and as per my previous comment, `LargeTitleViewTokenSet` is an unsuitable name.

If you absolutely insist on this being a separate PR, see #1686, but reverting it here just to immediately change it back after this PR gets merged into the feature branch is just plain silly.

File: ios/FluentUI/Navigation/NavigationBar.swift (hunk 1 above)
Comment (laminesm) [reply ios/FluentUI/Navigation/NavigationBar.swift:112]: The point of this function was to expose only the system style color of the nav bar. I was not aware of any situation in which we needed to access the colors from the other style while updating OMR to fluent 2. Also not sure how this counts as debt...

File: ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift (hunk 3 above)
Comment (amgleitman) [reply ios/FluentUI/Core/Theme/Tokens/ControlTokenSet.swift:63]: Using [this bit of code](https://github.com/microsoft/fluentui-apple/blob/b7e84b4137ab808239940b2b37892285ca5f41e9/ios/FluentUI/Tab%20Bar/SideTabBar.swift#L381) as a reference point, I think what we really want here is to replace the RHS with `otherTokenSet.overrideValue(forToken: otherToken)`.

File: ios/FluentUI/TwoLineTitleView/TwoLineTitleView.swift [hunk 10]
```diff
@@ -280,15 +278,18 @@ open class TwoLineTitleView: UIView, TokenizedControlInternal {
...
-        let subtitleIsNilOrEmpty = subtitle?.isEmpty ?? true
-        titleButtonLabel.maxFontSize = subtitleIsNilOrEmpty ? 0 : fluentTheme.aliasTokens.typography[.body1Strong].size
-
         setupButton(titleButton, label: titleButtonLabel, trailingImageView: titleButtonTrailingImageView, text: title, interactive: interactivePart.contains(.title), accessoryType: accessoryType)
         // Check for strict equality for the subtitle button's interactivity.
         // If the whole area is active, we'll stretch the title button to adjust the hit area
         // while still only keeping one button active from an accessibility standpoint.
         setupButton(subtitleButton, label: subtitleButtonLabel, trailingImageView: subtitleButtonImageView, text: subtitle, interactive: interactivePart == .subtitle, accessoryType: accessoryType)
 
+        if #available(iOS 15.0, *) {
```
Comment (mischreiber) [inline ios/FluentUI/TwoLineTitleView/TwoLineTitleView.swift:287]: This isn't necessary - the minimum SDK is already moved to 15.0
